metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -140,6 +144,11 @@ public PartitionChangeBuilder setTargetLeaderRecoveryState(LeaderRecoveryState t
         return this;
     }
 
+    public PartitionChangeBuilder setBumpLeaderEpochOnIsrShrink(boolean bumpLeaderEpochOnIsrShrink) {
+        this.bumpLeaderEpochOnIsrShrink = bumpLeaderEpochOnIsrShrink;
+        return this;
+    }
```

Comment:
Hmm. Doesn't this ignore the MV? Checking the MV is important to make sure that all of the brokers support this feature before the controller can skip leader epoch bumps. I think you want the following code:
```java
    public PartitionChangeBuilder setBumpLeaderEpochOnIsrShrink(boolean bumpLeaderEpochOnIsrShrink) {
        // If the MV doesn't support skipping leader epoch bump the caller cannot override it.
        if (metadataVersion.isSkipLeaderEpochBumpSupported()) {
           this.bumpLeaderEpochOnIsrShrink = bumpLeaderEpochOnIsrShrink;
        }
        return this;
    }
```

Because of this, it is not clear to me if there is even the right API. You probably want `setInZKMigrationMode` to make all of this more explicit.

Reply:
> Hmm. Doesn't this ignore the MV?

No. The MV is consulted in the constructor.

> Checking the MV is important to make sure that all of the brokers support this feature before the controller can skip leader epoch bumps

This function is only used to enable bumping leader epoch, which is always safe to do, even on older MVs.

If you are really concerned about foolproofing this you could make a PR that changes this function from `setBumpLeaderEpochOnIsrShrink` to `enableBumpLeaderEpochOnIsrShrink` (i.e. remove the option to disable this way)

Reply:
Oops, the intention here was to only call this API when ZK migration was enabled.

To fix this, I’d suggest removing setBumpLeaderEpochOnIsrShrink and changing it to enableBumpLeaderEpochOnIsrShrink. Basically, the API can be used to enable the bump, but it cannot be used to disable. This should avoid this kind of problem in the future.